src/contrasts.jl

```diff
@@ -466,49 +474,79 @@ expressed as a hypothesis matrix, but it is not obvious just from looking at the
 contrasts matrix.  For this reason `HypothesisCoding` is preferred for
 specifying custom contrast coding schemes over `ContrastsCoding`.
 
+Optional arguments `levels` and `labels` give the ordering of mapping between
+levels and columns of the hypothesis matrix, and the labels for the generated
+contrasts, respetively.
 """
-mutable struct HypothesisCoding <: AbstractContrasts
-    hypotheses::Matrix
-    contrasts::Matrix
-    base::Nothing
-    levels::Union{Vector,Nothing}
-
-    function HypothesisCoding(hypotheses, base, levels)
+mutable struct HypothesisCoding{T<:AbstractMatrix, S<:AbstractMatrix} <: AbstractContrasts
+    hypotheses::T
+    contrasts::S
+    levels::Union{AbstractVector,Nothing}
+    labels::AbstractVector
+
+    function HypothesisCoding(hypotheses::T, levels, labels) where {T}
```

Comment:
there are kwarg methods further down, this is just to ensure that you can't construct a HypothesisCoding with mismatching contrasts and hypothesis matrices...

Reply:
But couldn't there just be one single (inner) constructor taking keyword arguments?

Reply:
hm, maybe.  I hadn't thought about that.  I'm not sure we want to force people to use keyword arguments but honestly I don't have a strong opinion about it..

Reply:
In general AFAICT APIs choose either positional or keyword arguments, but don't support both.

Reply:
I see, and I agree that the kwargs are a better api.  But changing that would be a breaking change; there are a few contrasts-related breaking changes (keyword args, error on missing levels for hypothesis, type parameters) so maybe we should merge something that's non-breaking now and then make another breaking release.  Or just give up and make a breaking release next?  we're not on 1.0 yet so that's not SO disruptive I guess.

Reply:
Ah, right, I guess that's fine then. Maybe just leave an issue with the 1.0 milestone so that we decide whether to deprecate positional arguments at that point.

Reply:
1.0 seems a little ambitious at this point but I made a 0.7 :)